Lecture 02 (xgx / Guangxuan Xiao) (#8)
mit-han-lab/6s965-fall2022: notes/lecture-02-xgx.md

# Lecture 02: Basics of Neural Networks

## Note Information

> | Title       | Basics of Neural Networks                                                                     |
> | ----------- | --------------------------------------------------------------------------------------------- |
> | Lecturer    | Song Han                                                                                      |
> | Date        | 09/13/2022                                                                                    |
> | Note Author | Guangxuan Xiao (xgx)                                                                          |
> | Description | Reviewing the basics of deep learning and introducing efficiency metrics for neural networks. |

Reviewer: ```suggestion
| Title       | Basics of Neural Networks                                                                |
| ----------- | ---------------------------------------------------------------------------------------- |
| Lecturer    | Song Han                                                                                 |
| Date        | 09/13/2022                                                                               |
| Note Author | Guangxuan Xiao (xgx)                                                                     |
| Description | Review the basics of deep learning and introduce efficiency metrics for neural networks. |
```

> ## Terminology of Neural Networks

Reviewer: ```suggestion
## Terminology of neural networks
```

Synapses = Weights = Parameters: Connection intensities in neural nets.

> Neurons  = Features = Activations: Input / intermediate / Output values in neural nets.

Reviewer: ```suggestion
Neurons  = Features = Activations: Input / intermediate / output values in neural nets.
```

> ## Popular Neural Network Layers

Reviewer: ```suggestion
## Popular neural network layers
```

> ### Fully-connected layer (Linear layer)

Reviewer: ```suggestion
### Fully-connected layer (linear layer)
```

The output neuron is connected to all input neurons.

> $y_i = \sum_j w_{ij}x_j+b_i$
> 
> $n$: Batch size,
> 
> $c_i$: Input Channels
> 
> $c_o$: Output Channels

Reviewer: ```suggestion
$y_i = \sum_j w_{ij}x_j+b_i$, where $n$ is the batch size, $c_i$ is the number of input channels, and $c_o$ is the number of output channels.
```

| Tensors             | Shapes       |
| ------------------- | ------------ |
| Input Features $X$  | $(n, c_i)$  |
| Output Features $Y$ | $(n, c_o)$   |
| Weights $W$         | $(c_o, c_i)$ |

> | bias $b$            | $(c_o, )$    |

Reviewer: ```suggestion
| Bias $b$            | $(c_o, )$    |
```

> ### Convolution Layer

Reviewer: ```suggestion
### Convolution layer
```

> The output neuron is connected to input neurons in the receptive field.

Reviewer: ```suggestion
The output neuron is only connected to input neurons in the receptive field.
```

| Tensors             | Shapes of 1-D Conv | Shapes of 2-D Conv     |
| ------------------- | ------------------ | ---------------------- |

> | Input Features $X$  | $ (n, c_i, w_i)$   | $(n, c_i, h_i, w_i)$   |

Reviewer: Same here.

| Output Features $Y$ | $(n, c_o, w_o)$    | $(n, c_o, h_o, w_o)$   |
| Weights $W$         | $(c_o, c_i,k_w)$   | $(c_o, c_i, k_h, k_w)$ |

> | bias $b$            | $(c_o, )$          | $(c_o, )$              |

Reviewer: ```suggestion
| Bias $b$            | $(c_o, )$          | $(c_o, )$              |
```

- Multiplications needed to calculate a single output element equals the size of the convolution kernels.

- How to calculate output height and width?

> $h_o = h_i - k_h + 1$

Reviewer: Please wrap it into the previous line.

#### Padding

> Padding can keep the output feature map size the same as the  input feature map size.

Reviewer: ```suggestion
Padding keeps the output feature map size the same as the input feature map size.
```

- Zero padding pads the input boundaries with zero.
  

>   $h_o + h_i + 2p-k_h + 1$

Reviewer: ```suggestion
  $h_o = h_i + 2p-k_h + 1$
```

>   $h_0 + h_i + 2p-k_h + 1$

Reviewer: `$h_0 + h_i + 2p-k_h + 1$` -> $h_o = h_i + 2p-k_h + 1$

> - Other: Reflection Padding, replication padding, constant padding.

Reviewer: ```suggestion
- Other: reflection padding, replication padding, constant padding.
```

> #### Receptive Field

Reviewer: ```suggestion
#### Receptive field
```

> - In convolution , each output elelment depends on $k_h \times k_w$  in the input.

Reviewer: ```suggestion
- In convolution, each output element depends on $k_h \times k_w$ in the input.
```

> - Each successive convulution adds $k-1$ to the receptive field size.

Reviewer: ```suggestion
- Each successive convolution adds $k-1$ to the receptive field size.
```

- With L layers the receptive field size is $L (k-1) + 1$.

> #### Strided Convulution Layer

Reviewer: ```suggestion
#### Strided convolution layer
```

> Strided Convulution Layers increase the receptive field without increasing the layer.

Reviewer: ```suggestion
Strided convolution layers increase the receptive field without increasing the layer.
```

$h_o = \frac{h_i + 2p-k_h}{s}+1$, where s is the stride and p is the padding.

> #### Group Convolution Layer

Reviewer: ```suggestion
#### Group convolution layer
```

> A group of narrower convolutions.

Reviewer: This is not very clear. Please define with more details.

Reviewer: This is not resolved.

Author: Resolved

Weights W: $(c_o, c_i, k_h, k_w) \rightarrow (g \times c_o/g, c_i/g, k_h, k_w)$  

#### Depthwise convolution layer

Independent filter for each channel: $g= c_i = c_o$ in grouped convolution.

Weights: $(c, k_h, k_w)$

#### Other convolution layers

> - Transposed Convolution: super-resolution, segmentation, backprop.

Reviewer: ```suggestion
- Transposed convolution can be used to upsample the feature map (e.g., in super-resolution and segmentation).
```

> - Dilated Convolution

Reviewer: ```suggestion
- Dilated convolution can be used to increase the receptive field with dilated convolution kernels.
```

### Pooling layer

> Downsample the feature map to a smaller size, and graduallyreduceg the size of the feature maps.

Reviewer: ```suggestion
Pooling layers are used to downsample the feature map and gradually reduce the size of feature maps.
```

Usually, the stride is the same as the kernel size.

> ### Normalization Layer

Reviewer: ```suggestion
### Normalization layer
```

> Normalizing the features makes training faster

Reviewer: ```suggestion
Normalizing the features makes the model optimization faster.
```

- The normalization layer normalized the features as follows
  
  $\hat{x}_i = \frac{x_i-\mu_i}{\sigma}$,
  
  $\mu_i = \frac{1}{m}\sum_{i\in\delta_i}x_k$,
  
  $\sigma_i=\sqrt{\frac{1}{m}\sum_{k\in\delta_i}(x_k-\mu_i)^2+\varepsilon}$

### Activation function

> Non-linear functions

Reviewer: ```suggestion
Activation functions are usually non-linear.
```

> - most efficient: ReLU
> 
> - ReLU problem: input lower than zero will have 0 gradient
> 
> - Quantization friendly: ReLU

Reviewer: ```suggestion
- ReLU is efficient, and quantization-friendly, but has zero gradients for negative inputs.
```

> - GeLU, Swish, hard swish: hard to quantize

Reviewer: ```suggestion
- GeLU, Swish, and Hardswish are more effective but hard to be quantized.
```

### Transformers

> Attention: finding the relationship between each token, problem $O(n^2)$ complexity and when $n$ is very large it will be very slow.

Reviewer: ```suggestion
Transformer is composed of multiple self-attention layers, which find the relationship between tokens. The computational complexity of self-attention is $O(n^2)$, where $n$ is the token size. Therefore, it will be very slow with large token size.
```

> ## Popular CNN Architectures

Reviewer: ```suggestion
## Popular CNN architectures
```

> - AlexNet: the first architecture to adopt an architecture with consecutive convolutional layers (conv layer 3, 4, and 5).

Reviewer: ```suggestion
- AlexNet is the first to adopt an architecture with consecutive convolutional layers.
```

> - ResNet: introduce a residual connection to allow gradient flow.

Reviewer: ```suggestion
- ResNet introduces residual connections to allow better gradient flow.
```

> - MobileNetV2: depth-wise convolution

Reviewer: ```suggestion
- MobileNetV2 introduces depthwise convolution to effectively reduce the computation cost.
```

> - MobileNetV2: depth-wise convolution

Reviewer: Please add citations for these papers.

> ## Efficiency metrics for Deep learning

Reviewer: ```suggestion
## Efficiency metrics for deep learning
```

Core: Computation and memory

> Three Dimensions: Storage, Latency, and Energy

Reviewer: Please complete them as sentences.

### Latency

Latency = $max(T_{operation}, T_{memory})$

### Energy consumption

> Data movement - > more memory reference -> more energys

Reviewer: `- >` -> `->`

> DRAM Access > SRAM Access > FP Mult > INT Mult > Register File > FP Add > INT Add

Reviewer: Please complete them as sentences.

Reviewer: This is not resolved.

Author: Resolved

> ### Memory related

Reviewer: ```suggestion
### Memory-related metrics
```

> #### Number of Parameters

Reviewer: ```suggestion
#### Number of parameters
```

- Linear: $c_0 \times c_i$

- Convolution: $c_o c_i k_h k_w$

> - Grouped Convolution: $c_o c_i k_h k_w / g$

Reviewer: ```suggestion
- Grouped convolution: $c_o c_i k_h k_w / g$
```

> - Depth-wise convolution: $c_o k_w k_h$

Reviewer: ```suggestion
- Depthwise convolution: $c_o k_w k_h$
```

#### Model size

- $Model Size = NumOfParameters \times BitWidth$

> - e.g. AlexNet 61M parameters, using fp32 = 244MB, using int8 = 61MB

Reviewer: ```suggestion
For example, AlexNet has 61M parameters, so its model size would be 244MB (with FP32) and 61MB (with INT8).
```

> #### Total / Peak activation

Reviewer: ```suggestion
#### Total and peak activation
```

- Imbalanced memory distribution of MobileNet: the bottleneck is the peak (inference) and sum (training)

- Training bottleneck: not weight but activation

- MCUNet: Activation getting smaller, weights getting larger because the channel is increasing and the resolution is decreasing.

> ### Computation related

Reviewer: ```suggestion
### Computation-related metrics
```

> #### MACs: Multiply-Accumulate Operations

Reviewer: ```suggestion
#### MACs: multiply-accumulate operations
```

> MAC: $a = a+b\times c$

Reviewer: ```suggestion
A multiply-accumulate (MAC) operation is $a = a+b\times c$. Here are the number of MACs for common operators:
```

- Matrix-vector multiplication (MV): $m\times n$

- General matrix-matrix multiplication (GEMM): $m\times n\times k$

- Linear layer: $c_o\times c_i$

- Convolution: $c_i\times k_w \times k_h \times h_o \times w_o \times c_o$

> - Grouped Convolution: $c_i\times k_w \times k_h \times h_o \times w_o \times c_o / g$

Reviewer: ```suggestion
- Grouped convolution: $c_i\times k_w \times k_h \times h_o \times w_o \times c_o / g$
```

> - Depthwise Convolution: $k_w \times k_h \times h_o \times w_o \times c_o$

Reviewer: ```suggestion
- Depthwise convolution: $k_w \times k_h \times h_o \times w_o \times c_o$
```

> - AlexNet have 724MACs

Reviewer: ```suggestion
As an example, AlexNet has 724M MACs.
```

> #### FLOP: Floating Point Operation

Reviewer: ```suggestion
#### FLOP: floating point operation
```

> - One floating-point MAC = Two FLOPs

Reviewer: ```suggestion
One floating-point multiply-accumulate (MAC) operation corresponds to two floating-point operations (FLOP): multiply and add.
```

> e.g. AlexNet has 724 MACs, so the total number of floating point operations will be 1.4G FLOPs.

Reviewer: ```suggestion
For instance, AlexNet has 724M MACs, so the number of floating point operations will be 1.4G.
```

> - Floating Point Operation Per Second (FLOPS)

Reviewer: ```suggestion
- Floating point operation per second (FLOPS)
```
